XMonad/Layout/DecorationEx/LayoutModifier.hs

```diff
@@ -0,0 +1,322 @@
+          toadd a = filter (\p -> fst p `elem` a)
+
+          createDecoWindowIfNeeded :: Maybe Window -> Maybe Rectangle -> X (Maybe Window)
+          createDecoWindowIfNeeded mbDecoWindow mbDecoRect =
+            case (mbDecoWindow, mbDecoRect) of
+              (Nothing, Just decoRect) -> do
+                decoWindow <- createDecoWindow engine theme decoRect
+                return $ Just decoWindow
+              _ -> return mbDecoWindow
+
+          resync :: DecorationEngineState engine -> [WindowDecoration] -> [(Window,Rectangle)] -> X [WindowDecoration]
+          resync _ _ [] = return []
+          resync decoState dd ((window,rect):xs) =
+            case  window `elemIndex` getOrigWindows dd of
+              Just i  -> do
+                mbDecoRect <- decorateWindow geom screenRect stack srcPairs (window,rect)
+                widgetPlaces <- case mbDecoRect of
+                                  Nothing -> return $ WidgetLayout [] [] []
+                                  Just decoRect -> placeWidgets engine theme shrinker decoState decoRect window (themeWidgets theme)
+                mbDecoWindow  <- createDecoWindowIfNeeded (findDecoWindow i dd) mbDecoRect
+                let newDd = WindowDecoration window rect mbDecoWindow mbDecoRect (widgetLayout widgetPlaces)
+                restDd <- resync decoState dd xs
+                return $ newDd : restDd
+              Nothing -> resync decoState dd xs
+
+          -- We drop any windows that are *precisely* stacked underneath
+          -- another window: these must be intended to be tabbed!
+          removeTabbed :: [Rectangle] -> [(Window, Rectangle)] -> [(Window, Rectangle)]
+          removeTabbed _ [] = []
+          removeTabbed rs ((w,r):xs)
+              | r `elem` rs = removeTabbed rs xs
+              | otherwise   = (w,r) : removeTabbed (r:rs) xs
+
+          insertDwr :: WindowDecoration -> [(Window, Rectangle)] -> [(Window, Rectangle)]
+          insertDwr dd wrs =
+            case (wdDecoWindow dd, wdDecoRect dd) of
+              (Just decoWindow, Just decoRect) -> (decoWindow, decoRect) : (wdOrigWindow dd, shrinkWindow geom decoRect (wdOrigWinRect dd)) : wrs
+              _ -> (wdOrigWindow dd, wdOrigWinRect dd) : wrs
+
+          dwrs_to_wrs :: [WindowDecoration] -> [(Window, Rectangle)]
+          dwrs_to_wrs = removeTabbed [] . foldr insertDwr []
+
+          processState :: DecorationLayoutState engine -> X ([(Window, Rectangle)], Maybe (DecorationEx engine widget geom shrinker Window))
+          processState st = do
+            let decorations = dsDecorations st
+            showDecos decorations
+            updateDecos engine shrinker theme (dsStyleState st) decorations
+            return (dwrs_to_wrs decorations, Just (DecorationEx (I (Just (st {dsDecorations = decorations}))) shrinker theme engine geom))
+
+    handleMess (DecorationEx (I (Just st)) shrinker theme engine geom) m
+        | Just e <- fromMessage m = do
```

Comment:
Will will always shadow the other two matches, right? Since you're only matching on `fromMessage` anyways, I think using `case` here might be clearer (and will alert you to overlapping matches)

Reply:
Thanks, I did not see that I have fetch-all pattern first in the order. Will move it to the end.
As far as I understand, it's not possible to use `case` here, because different branches match against different types returned by `fromMessage`. Another pattern used in xmonad code is smth like
```
handleMess ... = msum [handleHide <$> fromMessage m, handleRelease <$> fromMessage m ...]
```
but I'm not sure if that is much more readable.

Reply:
> As far as I understand, it's not possible to use case here, because different branches match against different types returned by fromMessage.

Ah, quite right!